code_generator/GdkCodeGenerator/Templates/UnityGameObjectComponentDispatcherGenerator.tt

```diff
@@ -192,16 +234,80 @@ namespace <#= qualifiedNamespace #>
                         continue;
                     }
 
-                    var authChanges = authChangeLists[i];
-                    foreach (Requirables.ReaderWriterImpl readerWriter in readersWriters)
+                    if (isFirstAuthChange(Authority.Authoritative, changeOpsLists[i]))
                     {
-                        foreach (var auth in authChanges.Changes)
+                        foreach (Requirables.ReaderWriterImpl readerWriter in readersWriters)
                         {
-                            readerWriter.OnAuthorityChange(auth);
+                            readerWriter.OnAuthorityChange(Authority.Authoritative);
                         }
                     }
                 }
             }
+
+            public override void InvokeOnAuthorityLostCallbacks(Dictionary<int, InjectableStore> entityIndexToInjectableStore)
+            {
+                if (AuthorityLostComponentGroup.IsEmptyIgnoreFilter)
+                {
+                    return;
+                }
+
+                var entities = AuthorityLostComponentGroup.GetEntityArray();
+                var changeOpsLists = AuthorityLostComponentGroup.GetComponentDataArray<AuthorityChanges<<#= componentDetails.TypeName #>>>();
+
+                // Call once on all entities unless they flip-flopped back into the state they started in
```

Comment:
Semantically, wouldn't you want to invoke all expected callbacks, even if they ultimately change back? cc @samiwh @jamiebrynes7

Reply:
This is necessary with the approach we are taking to the order of passing ops and enabling behaviours in the MonoBehaviour layer - we are not retaining the information about the exact order of all ops received in the GDK Core, and as a result the MonoBehaviour layer performs all enables at a different time than all disables. So not squashing them could mean receiving multiple disables and then multiple enables for a component rather than alternating them.

Reply:
^
We could invoke all callbacks if we preserve the op order in the core and provide that to the monobehaviour layer. Which we can do if it becomes essential. But that comes with it's own disadvantages of making users think the state of the worker is something other than it is. Right now there is one race we haven't dealt with when you receive authority cycles however that can be fixed by adding to this later (wasn't considered essential for the first set of corrections).